src/data/fintype/card_embedding.lean

```diff
@@ -0,0 +1,156 @@
+Authors: Eric Rodriguez
+-/
+import data.fintype.card
+import data.equiv.fin
+import data.nat.factorial
+import tactic
+import tactic.slim_check
+
+/-!
+# Birthday Problem
+
+This file establishes the cardinality of `α ↪ β` in full generality.
+-/
+
+open finset
+
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+open_locale nat
+
+namespace fintype
+
+-- why does `ᶜ` not work here but elsewhere it does? aren't they defeq due to the instance?
+def to_be_named {α β γ : Type*} :
+((α ⊕ β) ↪ γ) ≃ (Σ f : α ↪ γ, β ↪ (set.range f).compl) :=
+{ to_fun := λ f, ⟨function.embedding.inl.trans f, -- α ↪ γ
+  ⟨λ b, ⟨f (sum.inr b), -- β → (f α)ᶜ
+
+    begin --proving membership in not the range
+      apply set.mem_compl,
+      suffices : ∀ (x : α), ¬f (sum.inl x) = f (sum.inr b), by simpa,
+      intros x,
+      rw [function.injective.eq_iff f.injective],
+      simp only [not_false_iff]
+    end⟩,
+
+  begin -- prove injective(β → (f α)ᶜ)
+    intros a b f_eq,
+    simp only [subtype.mk_eq_mk] at f_eq,
+    apply_fun @sum.inr α β using sum.inr_injective,
+    apply_fun f using function.embedding.injective f,
+    exact f_eq
+  end⟩⟩,
+
+  inv_fun := λ ⟨f, g⟩,
+    ⟨(equiv.sum_arrow_equiv_prod_arrow _ _ _).symm ⟨f, (λ x, g x)⟩, -- implicit coe here
+
+  begin -- prove that this amalgamation is injective; think this is the best way
+    rintros (a₁|b₁) (a₂|b₂) f_eq;
+    rw equiv.sum_arrow_equiv_prod_arrow at f_eq;
+    simp only [equiv.coe_fn_symm_mk, sum.elim_inl, sum.elim_inr] at f_eq,
+    { rw f.injective f_eq },
+    { exfalso, apply (g b₂).property, rw [subtype.val_eq_coe, ←f_eq], simp },
+    { exfalso, apply (g b₁).property, rw [subtype.val_eq_coe, f_eq], simp },
+    { rw g.injective (subtype.coe_injective f_eq) }
+  end⟩,
+
+  left_inv := λ f, by { ext, cases x; simp! [equiv.sum_arrow_equiv_prod_arrow] },
+
+  right_inv := λ ⟨g, h⟩, by begin
+    dsimp only,
+    have : { function.embedding . to_fun := ⇑h, inj' := h.injective } = h, by { ext, simp },
+    ext; simp! [equiv.sum_arrow_equiv_prod_arrow, this]
+  end }
+
+def some_other_nameless_thing {α : Type*} : (fin 1 ↪ α) ≃ α :=
+{ to_fun := λ f, f 0,
+  inv_fun := λ x, ⟨λ _, x, λ _ _ _, subsingleton.elim _ _⟩,
+  left_inv := λ _, by { ext, simp_rw [function.embedding.coe_fn_mk], congr },
+  right_inv := λ _, by simp }
+
+private lemma card_embedding_aux (n : ℕ) (β) [fintype β] [decidable_eq β] (h : n ≤ ‖β‖) :
+  ‖fin n ↪ β‖ = nat.desc_fac (‖β‖ - n) n :=
+begin
+  induction n with n hn,
+  { nontriviality (fin 0 ↪ β),
+    rw [nat.desc_fac_zero, fintype.card_eq_one_iff],
+    refine ⟨nonempty.some nontrivial.to_nonempty, λ x, function.embedding.ext fin.elim0⟩ },
+
+  rw [nat.succ_eq_add_one, ←card_congr (equiv.embedding_congr fin_sum_fin_equiv (equiv.refl β))],
+  rw card_congr to_be_named,
+  all_goals { try { apply_instance } },
+
+  have : ∀ (f : fin n ↪ β), ‖fin 1 ↪ ↥((set.range f).compl)‖ = ‖β‖ - n,
```

Comment:
This should certainly be a separate lemma, just computing `‖fin 1 ↪ X‖ = card X`.

Reply:
Or more generally, `‖T ↪ X‖ = card X` given `[unique T]`.

Reply:
didn't know about [`unique`] - thanks!

Reply:
okay, this lemma is cursed:

```
lemma card_embedding_of_unique {α β : Type*} [unique α] [fintype β] [decidable_eq α] [decidable_eq β] :
‖α ↪ β‖ = ‖β‖ := card_congr some_other_nameless_thing
```

If I even dare put it in the simpset, it causes a timeout (even if I comment out all simps apart from `nontriviality`! So it shouldn't even be looking there!). Trying to `rw` with it also causes Lean to give up, but even more weirdly, `card_congr some_other_nameless_thing` is just fine. Maybe it's the subsingleton instances again, but I'm really tempted not to put it in as it's toxic waste

Reply:
Maybe try adding `[fintype α]`?

Reply:
doesn't help :(

Reply:
What fintype instance is the statement using? Can you push the failing definition to a branch so I can try it out in context?

Another idea might be to use `[subsingleton α] [nonempty α]` instead.

Reply:
https://github.com/leanprover-community/mathlib/commit/3ce5e10a706e23d90feb32f505a84802307c39e0 commit with issues here

Reply:
What's the status here? Maybe you can start a topic on zulip to ask for help if necessary?

Reply:
I'll try bring it up to date, play around with it a little, and then ask in Zulip if not

Reply:
okay, with the new code, it _seems_ to work just fine locally! hopefully CI agrees